File: android/app/src/main/kotlin/com/goalmaker/app/domain/sync/SyncedTableCatalog.kt
package com.goalmaker.app.domain.sync

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.jsonPrimitive

/** The synced tables in push and pull order, read from contracts/schemas/synced-tables.json. */
class SyncedTableCatalog(
    /** Parents before children, the order pushes and pulls follow. */
    val tables: List<SyncedTable>,
) {
    private val byName = tables.associateBy(SyncedTable::name)

    operator fun get(name: String): SyncedTable =
        byName[name] ?: throw NoSuchElementException("'$name' is not a synced table")

    companion object {
        fun parse(json: String): SyncedTableCatalog {
            val tables = Json.parseToJsonElement(json).jsonObject.getValue("tables").jsonArray.map { table ->
                val fields = table.jsonObject
                SyncedTable(
                    name = fields.getValue("name").jsonPrimitive.content,
                    columns = fields.getValue("columns").jsonArray.map { column ->
                        SyncedColumn(
                            name = column.jsonObject.getValue("name").jsonPrimitive.content,
                            kind = ColumnKind.parse(column.jsonObject.getValue("kind").jsonPrimitive.content),
                            required = column.jsonObject["required"]?.jsonPrimitive?.booleanOrNull == true,
                        )
                    },
                )
            }
            return SyncedTableCatalog(tables)
        }
    }
}

File: windows/src/GoalMaker.Core/Sync/SyncedTableCatalog.cs
using System.Text.Json;

namespace GoalMaker.Core.Sync;

/// <summary>The synced tables in push and pull order, read from contracts/schemas/synced-tables.json.</summary>
public sealed class SyncedTableCatalog
{
    private readonly Dictionary<string, SyncedTable> byName;

    public SyncedTableCatalog(IReadOnlyList<SyncedTable> tables)
    {
        Tables = tables;
        byName = tables.ToDictionary(table => table.Name, StringComparer.Ordinal);
    }

    /// <summary>Parents before children, the order pushes and pulls follow.</summary>
    public IReadOnlyList<SyncedTable> Tables { get; }

    public SyncedTable this[string name] =>
        byName.TryGetValue(name, out var table) ? table : throw new KeyNotFoundException($"'{name}' is not a synced table");

    public static SyncedTableCatalog Load(Stream json)
    {
        using var document = JsonDocument.Parse(json);
        var tables = document.RootElement.GetProperty("tables").EnumerateArray()
            .Select(table => new SyncedTable(
                table.GetProperty("name").GetString()!,
                [.. table.GetProperty("columns").EnumerateArray().Select(column => new SyncedColumn(
                    column.GetProperty("name").GetString()!,
                    Enum.Parse<ColumnKind>(column.GetProperty("kind").GetString()!, ignoreCase: true),
                    column.TryGetProperty("required", out var required) && required.GetBoolean()))]))
            .ToList();
        return new SyncedTableCatalog(tables);
    }
}
